packages/SwingSet/test/test-controller.js

```diff
@@ -110,14 +110,41 @@ test('simple call', async t => {
   await simpleCall(t);
 });
 
-test('bootstrap', async t => {
+async function doTestBootstrap(t, doPin) {
   const config = await loadBasedir(
     new URL('basedir-controller-2', import.meta.url).pathname,
   );
+  if (doPin) {
+    config.pinBootstrap = true;
+  }
   const c = await buildVatController(config);
   t.teardown(c.shutdown);
   await c.run();
-  t.deepEqual(c.dump().log, ['buildRootObject called', 'bootstrap called']);
+  const kpid = c.queueToVatRoot('bootstrap', 'doMore');
+  await c.run();
+  const status = c.kpStatus(kpid);
+  const capdata = c.kpResolution(kpid);
+  if (doPin) {
+    t.deepEqual(c.dump().log, [
+      'buildRootObject called',
+      'bootstrap called',
+      'more stuff',
+    ]);
+  } else {
+    t.deepEqual(status, 'rejected');
+    t.deepEqual(capdata, {
+      body: '{"@qclass":"error","name":"Error","message":"vat terminated"}',
```

Comment:
We should be testing this by looking at the kernelDB for the bootstrap vat's c-list: in the `true` case, `o+0` should be present (and have a refcount of 1), in the `false` case it should be missing. `test/test-gc-kernel.js` has some helpers for this (like `dumpObjects()`). We want to figure out the kref, then use `t.deepEqual(dumpObjects()[kref], [bootstrapVatID, 1, 1])` for the `true` case.

The `vat terminated` message is kinda weird. Was it liveslots exploding when the kernel handed it a vref that wasn't present in `slotToVal`? That's the kernel's fault, and liveslots can't do anything better than explode (if it can't trust the kernel to maintain an invariant, it can't trust it with anything else, and without the kernel it can't do anything). But it's not really what this test wants to be asserting.

Sending a message with `c.queueToVatRoot` is a good additional check. I'd be inclined to only do it in the `true` case, but maybe I'm just trying to keep spurious errors out of the stdout/stderr stream.

Reply:
I think I disagree.  The purpose of the config option is to prevent the root object from being GC'd.  The fact that it does this by means of incrementing the refcount is secondary.  The test as implemented experiences the GC or not according to the option setting, which is exactly what we want to check.  There's already so much crap in the stdout/stderr stream that I think that ship has sailed.